File: 0547-number-of-provinces/05470-number-of-provinces.kt
class Solution {
    private val parent = IntArray(200) { it }

    fun findCircleNum(isConnected: Array<IntArray>): Int {
        val n = isConnected.size

        for (i in 0 until n) {
            for (j in 0 until n) {
                if (isConnected[i][j] == 1) makeUnion(i, j)
            }
        }
        
        val result = HashSet<Int>()
        for (i in 0 until n) {
            result.add(getParent(i))
        }
        return result.size
    }

    private fun getParent(x: Int): Int = if (parent[x] == x) x else getParent(parent[x])

    private fun makeUnion(x: Int, y: Int) {
        val nx = getParent(x)
        val ny = getParent(y)
        when (nx <= ny) {
            true -> parent[ny] = parent[nx]
            else -> parent[nx] = parent[ny]
        }
    }
}

File: 0547-number-of-provinces/05470-number-of-provinces.cpp
class Solution {
private:
    int parent[200];

    int getParent(int x) {
        if (x == parent[x]) return x;
        else return getParent(parent[x]);
    }

    void makeUnion(int x, int y) {
        x = getParent(x);
        y = getParent(y);
        if (x <= y) parent[y] = x;
        else parent[x] = y;
    }

public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        for (int i = 0; i < n; i++) parent[i] = i;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (isConnected[i][j]) makeUnion(i, j);
            }
        }

        unordered_set<int> result;
        for (int i = 0; i < n; i++) {
            result.insert(getParent(i));
        }
        return result.size();
    }
};

